### DISCUSSION TITLE
Overlap Grid Items?

### DISCUSSION THREAD

**Discussion by mirceaciu** [react-grid-layout/react-grid-layout#1421] [Ideas]
There are a few requests for overlapping elements #1062, #677, #373 and even some proposals to how to implement it #681.

Current project allows basic layouts to be created but for more advanced views like `picture-in-picture` it's not quite up to par. That's were overlapping elements could help.
 
As @fedwu has found out, making elements overlap is not hard. Just bypass this function https://github.com/react-grid-layout/react-grid-layout/blob/master/lib/utils.js#L197 or make it return `false`.

That above would allow elements to overlap but to make it an useful feature we would need a way to control the z-index of grid items. 

As an implementation:

- a flag on the `Grid Layout` to tell it to not check for collisions between grid items.
- two new control buttons `+` `-`  to increase / decrease the z-index of the target Grid Item. buttons would be similar to the 'drag to resize' icons/elements.

Z-index control is discussed in #858 (tl;dr: z-index needs to be added on both Grid Layout and Grid Items).    
If the items are read from an array the `+` `-` buttons can bind to functions that shifts the position of current element in that array and this should control the overlapping priority without setting specific z-indexes.  
 
Could any of this work?

**Comment by mirceaciu** [comment]
Small update. The flag on Grid Layout to disable the collision should be enough.

The zIndex does not seem to be an issue: it can be controlled with a  basic  "onClick" function.

In this next example I have:

- manually disabled collision check
- to each draggable Item I've attached a couple of buttons to increase or decrease the value of zIndex that I render inline with each element.

```
<GridLayout className="layout" 
    layout={layout}
    cols={10} 
    rowHeight={80} 
    width={800} 
    margin={[0, 0]} 
    verticalCompact={false} 
    isBounded={true}
    maxRows={5}
    onDrop={onDrop}
    isDroppable={true}
    droppingItem= {{ i: `wdgt_${index}`, w: 1, h: 1 }}
>
    {widgets.map((tile, index) => (
        <div className="widget" key={tile.i} data-grid={{isBounded: true}} style={{backgroundColor: tile.color, color: 'white', zIndex: tile.zIndex}}>
            {tile.i} {tile.zIndex}
            <button onClick={() => setZindex(index, 1)}> ++ </button>
            <button onClick={() => setZindex(index, -1)}> -- </button>
        </div>
    ))}
```

```
const [widgets, setWidgets] = useState([
    {i: "a", color: 'red', zIndex: 2},
    {i: "b", color: 'blue', zIndex: 1}
]);

const setZindex = (item_index, update_value) => {
    let current_widgets = widgets;

    current_widgets[item_index].zIndex = current_widgets[item_index].zIndex + update_value;
    setWidgets([...current_widgets]);
}
```

![zindex](https://user-images.githubusercontent.com/9830334/111527804-ffa0aa00-8768-11eb-81fd-c8b886d094e4.gif)

**Comment by mirceaciu** [comment]
also also: items can still be dragged and resized.

No sure if I can start a PR that adds the new flag but I can try :D

In this example I use `onMouseEnter` and `onMouseLeave` on the two buttons to disable or enable `isDraggable` prop on the `GridLayout`. That way clicking on the buttons won't start the drag event on the draggable element.

![draggable-and-indexable](https://user-images.githubusercontent.com/9830334/111535988-6080b000-8772-11eb-98a3-4fa533fbf29e.gif)

**Comment by mirceaciu** [comment]
found out there is a `preventCollision` prop but this only cancels the move or moves the overlapping elements so they don't collide.

I don't know if adding a new prop to the GridItems, 'allowOverlap', would break any of the current functionalities in unexpected ways
With this new prop we can ignore collisions like:

```
/**
 * Given two layoutitems, check if they collide.
 */
export function collides(l1: LayoutItem, l2: LayoutItem): boolean {
  
  // new check based on boolean flag allowOverlap
  if (l1.allowOverlap || l1.allowOverlap) return false // allow overlaps;

  if (l1.i === l2.i) return false; // same element
  if (l1.x + l1.w <= l2.x) return false; // l1 is left of l2
  if (l1.x >= l2.x + l2.w) return false; // l1 is right of l2
  if (l1.y + l1.h <= l2.y) return false; // l1 is above l2
  if (l1.y >= l2.y + l2.h) return false; // l1 is below l2
  return true; // boxes overlap
}
```

**Reply by nekw** [reply]
When will this feature be ready? Seeing an example 19 testing this. Looking forward to this feature!

**Reply by sehilyi** [reply]
Looking forward to this feature as well!

**Comment by klifa90** [comment]
is this or  #1470  being released to npm soon?

**Reply by sehilyi** [reply]
https://github.com/react-grid-layout/react-grid-layout/issues/1491

**Comment by mirceaciu** [comment]
@nekw , @sehilyi , @klifa90 

I lack the skills to create a proper PR, most I could do is document an adaptation I used in my local environment. I started this thread in the hopes that it's activity (your reactions) will motivate the author of this package to consider such a feature.

**Comment by booleanBoy** [comment]
Hoping this helps someone who lands here looking for ways to control collisions and overlapping widgets:

```
<ReactGridLayout
...
compactor={getCompactor(null, true, false)}
/>
```
The three params result in:
- No Compaction
- Allow Overlap
- Ignore Collisions
